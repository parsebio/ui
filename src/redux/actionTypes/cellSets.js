const CELL_SETS = 'cellSets';

/**
 * Updates the cells sets shown to the user with the hierarchy fetched
 * from the API.
 */
const CELL_SETS_LOADED = `${CELL_SETS}/loaded`;

/**
 * Signals to the UI that new cell set data is about to be loaded in.
 */
const CELL_SETS_LOADING = `${CELL_SETS}/loading`;

/**
 * Creates a new cell set.
 */
const CELL_SETS_CREATE = `${CELL_SETS}/create`;

/**
 * Updates a single cell set's properties.
 */
const CELL_SETS_UPDATE_PROPERTY = `${CELL_SETS}/updateProperty`;

/**
 * Cell sets were reordered
 */
const CELL_SETS_REORDER = `${CELL_SETS}/reorder`;

/**
 * Deletes a cell set.
 */
const CELL_SETS_DELETE = `${CELL_SETS}/delete`;

/**
 * Deletes a cell class.
 */
const CELL_CLASS_DELETE = `${CELL_SETS}/deleteClass`;

/**
 * Saves all cell sets currently loaded.
 */
const CELL_SETS_SAVE = `${CELL_SETS}/save`;

/**
 * Hide a cell set from visualizations.
 */
const CELL_SETS_HIDE = `${CELL_SETS}/hide`;

/**
 * Unhide [remove a previous hidden state] from a cell set from visualizations.
 */
const CELL_SETS_UNHIDE = `${CELL_SETS}/unhide`;

/**
 * Unhide all hidden cell sets.
 */
const CELL_SETS_UNHIDE_ALL = `${CELL_SETS}/unhideAll`;

// /**
//  * Cell sets clustering being recalculated.
//  */
const CELL_SETS_CLUSTERING_UPDATING = `${CELL_SETS}/clusteringUpdating`;

// /**
//  * Cell sets clustering being recalculated.
//  */
const CELL_SETS_CLUSTERING_UPDATED = `${CELL_SETS}/clusteringUpdated`;

/**
 * Creates an error condition in the cell set tool.
 */
const CELL_SETS_ERROR = `${CELL_SETS}/error`;

export {
  CELL_SETS_LOADING, CELL_SETS_LOADED,
  CELL_SETS_CREATE,
  CELL_SETS_UPDATE_PROPERTY, CELL_SETS_REORDER,
  CELL_SETS_DELETE,
  CELL_CLASS_DELETE,
  CELL_SETS_SAVE,
  CELL_SETS_HIDE, CELL_SETS_UNHIDE, CELL_SETS_UNHIDE_ALL,
  CELL_SETS_CLUSTERING_UPDATING, CELL_SETS_CLUSTERING_UPDATED,
  CELL_SETS_ERROR,
};
