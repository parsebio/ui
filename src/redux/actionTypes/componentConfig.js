/**
 * Sets plot configs and data.
 */
const COMPONENT_CONFIG = 'componentConfig';
const LOAD_CONFIG = `${COMPONENT_CONFIG}/load`;
const UPDATE_CONFIG = `${COMPONENT_CONFIG}/update`;
const RESET_CONFIG = `${COMPONENT_CONFIG}/reset`;
const SAVE_CONFIG = `${COMPONENT_CONFIG}/save`;
const CONFIGS_REPLACED = `${COMPONENT_CONFIG}/replacements`;
const PLOT_DATA_LOADED = `${COMPONENT_CONFIG}/plotDataLoaded`;
const PLOT_DATA_LOADING = `${COMPONENT_CONFIG}/plotDataLoading`;
const PLOT_DATA_ERROR = `${COMPONENT_CONFIG}/plotDataError`;
const PLOT_DATA_CLEARED = `${COMPONENT_CONFIG}/plotDataCleared`;

const TRAJECTORY_NODES_SELECTION_UPDATED = `${COMPONENT_CONFIG}/trajectoryNodesUpdated`;

export {
  LOAD_CONFIG,
  UPDATE_CONFIG, CONFIGS_REPLACED,
  RESET_CONFIG, SAVE_CONFIG,
  PLOT_DATA_LOADED, PLOT_DATA_LOADING, PLOT_DATA_ERROR,
  TRAJECTORY_NODES_SELECTION_UPDATED, PLOT_DATA_CLEARED,
};
