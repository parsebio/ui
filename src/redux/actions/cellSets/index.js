import createCellSet from './createCellSet';
import deleteCellSet from './deleteCellSet';
import deleteCellClass from './deleteCellClass';
import runCellSetsClustering from './runCellSetsClustering';
import runCellSetsAnnotation from './runCellSetsAnnotation';
import updateCellSetsClustering from './updateCellSetsClustering';

import loadCellSets from './loadCellSets';
import reorderCellSet from './reorderCellSet';
import updateCellSetProperty from './updateCellSetProperty';
import setCellSetHiddenStatus from './setCellSetHiddenStatus';
import unhideAllCellSets from './unhideAllCellSets';

export {
  createCellSet,
  deleteCellSet,
  deleteCellClass,
  runCellSetsClustering,
  runCellSetsAnnotation,
  updateCellSetsClustering,
  loadCellSets,
  reorderCellSet,
  updateCellSetProperty,
  setCellSetHiddenStatus,
  unhideAllCellSets,
};
