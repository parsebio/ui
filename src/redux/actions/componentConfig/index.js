import loadPlotConfig from 'redux/actions/componentConfig/loadPlotConfig';
import loadComponentConfig from 'redux/actions/componentConfig/loadComponentConfig';
import updatePlotConfig from 'redux/actions/componentConfig/updatePlotConfig';
import replaceLoadedConfigs from 'redux/actions/componentConfig/replaceLoadedConfigs';
import savePlotConfig from 'redux/actions/componentConfig/savePlotConfig';
import updatePlotData from 'redux/actions/componentConfig/updatePlotData';
import resetPlotConfig from 'redux/actions/componentConfig/resetPlotConfig';
import getDotPlot from 'redux/actions/componentConfig/getDotPlot';
import downloadNormalizedMatrix from 'redux/actions/componentConfig/downloadNormalizedMatrix';
import clearPlotData from 'redux/actions/componentConfig/clearPlotData';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  replaceLoadedConfigs,
  savePlotConfig,
  updatePlotData,
  resetPlotConfig,
  getDotPlot,
  downloadNormalizedMatrix,
  clearPlotData,
};
