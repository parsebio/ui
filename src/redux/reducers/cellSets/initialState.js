const initialState = {
  properties: {},
  hierarchy: [],
  initialLoadPending: true,
  loading: false,
  error: false,
  updatingClustering: false,
  hidden: new Set(),
};

export default initialState;
