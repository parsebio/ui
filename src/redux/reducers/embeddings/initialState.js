/**
 * The store has the following format:
 * {
 *    embeddingType: {
 *      data: {xValues:[1,2], yValues:[5,3], cellIds:[0,3]},
 *      loading: false,
 *      error: false
 *    }
 * }
 */

const initialEmbeddingState = {
  data: {},
  loading: true,
  error: false,
};

const initialState = {};

export { initialEmbeddingState };
export default initialState;
