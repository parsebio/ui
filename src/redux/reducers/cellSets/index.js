import {
  CELL_SETS_LOADING, CELL_SETS_LOADED,
  CELL_SETS_CREATE,
  CELL_SETS_UPDATE_PROPERTY,
  CELL_SETS_DELETE, CELL_CLASS_DELETE,
  CELL_SETS_HIDE, CELL_SETS_UNHIDE, CELL_SETS_UNHIDE_ALL,
  CELL_SETS_ERROR,
  CELL_SETS_CLUSTERING_UPDATING, CELL_SETS_CLUSTERING_UPDATED, CELL_SETS_REORDER,
} from 'redux/actionTypes/cellSets';
import { EXPERIMENT_SETTINGS_QC_START } from 'redux/actionTypes/experimentSettings';

import initialState from 'redux/reducers/cellSets/initialState';

import cellSetsLoading from 'redux/reducers/cellSets/cellSetsLoading';
import cellSetsLoaded from 'redux/reducers/cellSets/cellSetsLoaded';
import cellSetsUpdateProperty from 'redux/reducers/cellSets/cellSetsUpdateProperty';
import cellSetsDelete from 'redux/reducers/cellSets/cellSetsDelete';
import cellClassDelete from 'redux/reducers/cellSets/cellClassDelete';
import cellSetsReorder from 'redux/reducers/cellSets/cellSetsReorder';
import cellSetsCreate from 'redux/reducers/cellSets/cellSetsCreate';
import cellSetsError from 'redux/reducers/cellSets/cellSetsError';
import { cellSetsHide, cellSetsUnhide, cellSetsUnhideAll } from 'redux/reducers/cellSets/cellSetsHideUnhide';
import cellSetsClusteringUpdating from 'redux/reducers/cellSets/cellSetsClusteringUpdating';
import cellSetsClusteringUpdated from 'redux/reducers/cellSets/cellSetsClusteringUpdated';

const cellSetsReducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case CELL_SETS_LOADING: {
      return cellSetsLoading(state, action);
    }
    case CELL_SETS_LOADED: {
      return cellSetsLoaded(state, action);
    }
    case CELL_SETS_UPDATE_PROPERTY: {
      return cellSetsUpdateProperty(state, action);
    }
    case CELL_SETS_DELETE: {
      return cellSetsDelete(state, action);
    }
    case CELL_CLASS_DELETE: {
      return cellClassDelete(state, action);
    }
    case CELL_SETS_REORDER: {
      return cellSetsReorder(state, action);
    }
    case CELL_SETS_CREATE: {
      return cellSetsCreate(state, action);
    }
    case CELL_SETS_HIDE: {
      return cellSetsHide(state, action);
    }
    case CELL_SETS_UNHIDE: {
      return cellSetsUnhide(state, action);
    }
    case CELL_SETS_UNHIDE_ALL: {
      return cellSetsUnhideAll(state, action);
    }
    case CELL_SETS_CLUSTERING_UPDATING: {
      return cellSetsClusteringUpdating(state, action);
    }
    case CELL_SETS_CLUSTERING_UPDATED: {
      return cellSetsClusteringUpdated(state, action);
    }
    case CELL_SETS_ERROR: {
      return cellSetsError(state, action);
    }
    case EXPERIMENT_SETTINGS_QC_START: {
      return initialState;
    }
    default: {
      return state;
    }
  }
};

export default cellSetsReducer;
