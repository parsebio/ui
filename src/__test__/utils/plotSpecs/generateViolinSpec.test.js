import _ from 'lodash';
import SparseMap from 'utils/SparseMap';
import { generateData, generateSpec } from 'utils/plotSpecs/generateViolinSpec';
import { mockCellSets as cellSetsRaw } from '__test__/test-utils/cellSets.mock';
import { initialPlotConfigStates } from 'redux/reducers/componentConfig/initialState';
import getCellSets from 'redux/selectors/cellSets/getCellSets';

describe('generateData', () => {
  const MOCK_RANDOM = 0.1;
  const MOCK_EXPECTED_RANDOM = 0.25 + MOCK_RANDOM / 2;

  const expressionArray = [[0, 0], [1, 1], [2, 2], [3, 3], [4, 4], [5, 5], [6, 6]];
  const mockGeneExpression = new SparseMap(expressionArray);

  beforeEach(() => {
    jest.spyOn(global.Math, 'random').mockReturnValue(MOCK_RANDOM);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const cellSets = getCellSets()(cellSetsRaw);

  it('generates data when grouping by louvain clusters', () => {
    const groupingId = 'louvain';
    const plotData = generateData(cellSets, mockGeneExpression, groupingId, 'All');

    expect(_.keys(plotData.groups).sort()).toEqual(['cluster-a', 'cluster-b', 'cluster-c']);
    expect(plotData.groups['cluster-a']).toEqual({ name: 'cluster a', color: '#01FFFF' });

    const expectedCells = [
      { group: 'cluster-a', y: 0, x: MOCK_EXPECTED_RANDOM },
      { group: 'cluster-a', y: 1, x: MOCK_EXPECTED_RANDOM },
      { group: 'cluster-a', y: 6, x: MOCK_EXPECTED_RANDOM },
      { group: 'cluster-b', y: 2, x: MOCK_EXPECTED_RANDOM },
      { group: 'cluster-b', y: 3, x: MOCK_EXPECTED_RANDOM },
      { group: 'cluster-c', y: 4, x: MOCK_EXPECTED_RANDOM },
      { group: 'cluster-c', y: 5, x: MOCK_EXPECTED_RANDOM },
    ];

    expect(plotData.cells).toEqual(expectedCells);
  });

  it('generates data when grouping by sample', () => {
    const groupingId = 'sample';
    const plotData = generateData(cellSets, mockGeneExpression, groupingId, 'All');

    expect(_.keys(plotData.groups).sort()).toEqual(['sample-1', 'sample-2']);
    expect(plotData.groups['sample-1']).toEqual({ name: 'Sample 1', color: '#012FFF' });

    const expectedCells = [
      { group: 'sample-1', y: 0, x: MOCK_EXPECTED_RANDOM },
      { group: 'sample-1', y: 1, x: MOCK_EXPECTED_RANDOM },
      { group: 'sample-1', y: 2, x: MOCK_EXPECTED_RANDOM },
      { group: 'sample-2', y: 3, x: MOCK_EXPECTED_RANDOM },
      { group: 'sample-2', y: 4, x: MOCK_EXPECTED_RANDOM },
      { group: 'sample-2', y: 5, x: MOCK_EXPECTED_RANDOM },
    ];
    expect(plotData.cells).toEqual(expectedCells);
  });

  it('generates data when grouping by scratchpad', () => {
    const groupingId = 'scratchpad';
    const plotData = generateData(cellSets, mockGeneExpression, groupingId, 'All');

    expect(_.keys(plotData.groups).sort()).toEqual(['scratchpad-a']);
    expect(plotData.groups['scratchpad-a']).toEqual({ name: 'New Cluster', color: '#5FFFFF' });

    const expectedCells = [
      { group: 'scratchpad-a', y: 5, x: MOCK_EXPECTED_RANDOM },
    ];
    expect(plotData.cells).toEqual(expectedCells);
  });

  it('generates the x value for only the cells in the specified group', () => {
    const groupingId = 'louvain';
    const plotData = generateData(cellSets, mockGeneExpression, groupingId, 'sample/sample-1');
    const expectedCells = [
      { group: 'cluster-a', y: 0, x: MOCK_EXPECTED_RANDOM },
      { group: 'cluster-a', y: 1, x: MOCK_EXPECTED_RANDOM },
      { group: 'cluster-b', y: 2, x: MOCK_EXPECTED_RANDOM },
    ];
    expect(plotData.cells).toEqual(expectedCells);
  });

  it('generate spec returns the right data', () => {
    const data = {
      groups: {
        '8297d513-6268-4ab0-a03e-e74a23ecec07': {
          name: 'New Cluster',
          color: '#3957ff',
        },
        '99a7746b-7f9d-4e54-8acf-53031e4ff023': {
          name: 'New Cluster2',
          color: '#d3fe14',
        },
        '25ca1d7f-40ac-4bdc-9625-2272478e7db7': {
          name: 'New Cluster3',
          color: '#c9080a',
        },
        'b452c7ed-d907-4923-b7ff-fcafadc0b807': {
          name: 'New Cluster4',
          color: '#fec7f8',
        },
      },
      cells: [{
        group: '8297d513-6268-4ab0-a03e-e74a23ecec07',
        y: -0.1578865578865579,
        x: 0.5561686241217735,
      },
      {
        group: '8297d513-6268-4ab0-a03e-e74a23ecec07',
        y: -0.1578865578865579,
        x: 0.5683293645580038,
      },
      {
        group: '8297d513-6268-4ab0-a03e-e74a23ecec07',
        y: -0.1578865578865579,
        x: 0.4433627345700756,
      }],
      settings: {
        groupingName: 'Custom cell sets',
      },
    };
    const config = initialPlotConfigStates.violin;
    config.legend.enabled = true;
    const spec = generateSpec(config, data);
    expect(spec).toMatchSnapshot();
  });
});
