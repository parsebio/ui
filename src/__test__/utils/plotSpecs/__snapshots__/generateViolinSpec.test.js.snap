// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateData generate spec returns the right data 1`] = `
{
  "$schema": "https://vega.github.io/schema/vega/v5.json",
  "autosize": {
    "resize": true,
    "type": "fit",
  },
  "axes": [
    {
      "gridColor": "#000000",
      "gridOpacity": 0,
      "gridWidth": 0.5,
      "labelColor": "#000000",
      "labelFont": "sans-serif",
      "labelFontSize": 12,
      "offset": 10,
      "orient": "left",
      "scale": "yscale",
      "tickColor": "#000000",
      "title": "Z-Score of Expression",
      "titleColor": "#000000",
      "titleFont": "sans-serif",
      "titleFontSize": 13,
      "titlePadding": 5,
      "zindex": 1,
    },
    {
      "encode": {
        "labels": {
          "update": {
            "text": {
              "signal": "datum.value? data('groupCfg')[0][datum.value].name : ''",
            },
          },
        },
      },
      "gridColor": "#000000",
      "gridOpacity": 0,
      "gridWidth": 0.5,
      "labelAlign": "left",
      "labelAngle": 45,
      "labelColor": "#000000",
      "labelFont": "sans-serif",
      "labelFontSize": 12,
      "offset": 10,
      "orient": "bottom",
      "scale": "layout",
      "tickColor": "#000000",
      "title": "Custom cell sets",
      "titleColor": "#000000",
      "titleFont": "sans-serif",
      "titleFontSize": 13,
      "titlePadding": 5,
      "zindex": 1,
    },
  ],
  "background": "#FFFFFF",
  "data": [
    {
      "format": {
        "copy": true,
        "type": "json",
      },
      "name": "groupCfg",
      "values": {
        "25ca1d7f-40ac-4bdc-9625-2272478e7db7": {
          "color": "#c9080a",
          "name": "New Cluster3",
        },
        "8297d513-6268-4ab0-a03e-e74a23ecec07": {
          "color": "#3957ff",
          "name": "New Cluster",
        },
        "99a7746b-7f9d-4e54-8acf-53031e4ff023": {
          "color": "#d3fe14",
          "name": "New Cluster2",
        },
        "b452c7ed-d907-4923-b7ff-fcafadc0b807": {
          "color": "#fec7f8",
          "name": "New Cluster4",
        },
      },
    },
    {
      "format": {
        "copy": true,
        "type": "json",
      },
      "name": "cells",
      "values": [
        {
          "group": "8297d513-6268-4ab0-a03e-e74a23ecec07",
          "x": 0.5561686241217735,
          "y": -0.1578865578865579,
        },
        {
          "group": "8297d513-6268-4ab0-a03e-e74a23ecec07",
          "x": 0.5683293645580038,
          "y": -0.1578865578865579,
        },
        {
          "group": "8297d513-6268-4ab0-a03e-e74a23ecec07",
          "x": 0.4433627345700756,
          "y": -0.1578865578865579,
        },
      ],
    },
    {
      "name": "density",
      "source": "cells",
      "transform": [
        {
          "bandwidth": {
            "signal": "bandwidth",
          },
          "field": "y",
          "groupby": [
            "group",
          ],
          "type": "kde",
        },
      ],
    },
    {
      "name": "stats",
      "source": "cells",
      "transform": [
        {
          "as": [
            "q1",
            "median",
            "q3",
          ],
          "fields": [
            "y",
            "y",
            "y",
          ],
          "groupby": [
            "group",
          ],
          "ops": [
            "q1",
            "median",
            "q3",
          ],
          "type": "aggregate",
        },
      ],
    },
  ],
  "description": "Violin plot",
  "height": 550,
  "legends": [
    {
      "columns": 1,
      "direction": "horizontal",
      "encode": {
        "labels": {
          "update": {
            "fill": {
              "value": "#000000",
            },
            "text": {
              "field": "label",
              "scale": "sampleToName",
            },
          },
        },
      },
      "fill": "legend",
      "labelFont": "sans-serif",
      "labelLimit": 0,
      "offset": 40,
      "orient": "right",
      "symbolSize": 100,
      "symbolType": "circle",
      "type": "symbol",
    },
  ],
  "marks": [
    {
      "clip": true,
      "data": [
        {
          "name": "summaryOfGroup",
          "source": "stats",
          "transform": [
            {
              "expr": "datum.group === parent.group",
              "type": "filter",
            },
          ],
        },
        {
          "name": "cellsInGroup",
          "source": "cells",
          "transform": [
            {
              "expr": "datum.group === parent.group && isDefined(datum.x)",
              "type": "filter",
            },
          ],
        },
      ],
      "encode": {
        "enter": {
          "height": {
            "signal": "height",
          },
          "width": {
            "signal": "plotWidth",
          },
          "xc": {
            "band": 0.5,
            "field": "group",
            "scale": "layout",
          },
        },
      },
      "from": {
        "facet": {
          "data": "density",
          "groupby": "group",
          "name": "violin",
        },
      },
      "marks": [
        {
          "encode": {
            "enter": {
              "fill": {
                "signal": "data('groupCfg')[0][parent.group].color",
              },
              "orient": {
                "value": "horizontal",
              },
            },
            "update": {
              "width": {
                "field": "density",
                "scale": "xdensity",
              },
              "xc": {
                "signal": "plotWidth / 2",
              },
              "y": {
                "field": "value",
                "scale": "yscale",
              },
            },
          },
          "from": {
            "data": "violin",
          },
          "orient": "horizontal",
          "type": "area",
        },
        {
          "encode": {
            "update": {
              "fill": {
                "value": "black",
              },
              "fillOpacity": {
                "value": 0.5,
              },
              "opacity": {
                "value": 1,
              },
              "shape": {
                "value": "circle",
              },
              "size": {
                "value": 5,
              },
              "strokeWidth": {
                "value": 0,
              },
              "x": {
                "field": "x",
                "offset": {
                  "signal": "plotWidth / 2",
                },
                "scale": "xrandom",
              },
              "y": {
                "field": "y",
                "scale": "yscale",
              },
            },
          },
          "from": {
            "data": "cellsInGroup",
          },
          "type": "symbol",
        },
      ],
      "type": "group",
    },
  ],
  "padding": 5,
  "scales": [
    {
      "domain": {
        "data": "cells",
        "field": "group",
      },
      "name": "layout",
      "range": "width",
      "type": "band",
    },
    {
      "domain": {
        "data": "cells",
        "field": "y",
      },
      "name": "yscale",
      "nice": true,
      "range": "height",
      "type": "linear",
    },
    {
      "domain": {
        "data": "density",
        "field": "density",
      },
      "name": "xdensity",
      "range": [
        0,
        {
          "signal": "plotWidth",
        },
      ],
      "type": "linear",
    },
    {
      "domain": [
        0,
        1,
      ],
      "name": "xrandom",
      "range": [
        {
          "signal": "-plotWidth / 2",
        },
        {
          "signal": "plotWidth / 2",
        },
      ],
      "type": "linear",
    },
    {
      "domain": [
        "8297d513-6268-4ab0-a03e-e74a23ecec07",
        "99a7746b-7f9d-4e54-8acf-53031e4ff023",
        "25ca1d7f-40ac-4bdc-9625-2272478e7db7",
        "b452c7ed-d907-4923-b7ff-fcafadc0b807",
      ],
      "name": "legend",
      "range": [
        "#3957ff",
        "#d3fe14",
        "#c9080a",
        "#fec7f8",
      ],
      "type": "ordinal",
    },
    {
      "domain": [
        "8297d513-6268-4ab0-a03e-e74a23ecec07",
        "99a7746b-7f9d-4e54-8acf-53031e4ff023",
        "25ca1d7f-40ac-4bdc-9625-2272478e7db7",
        "b452c7ed-d907-4923-b7ff-fcafadc0b807",
      ],
      "name": "sampleToName",
      "range": [
        "New Cluster",
        "New Cluster2",
        "New Cluster3",
        "New Cluster4",
      ],
      "type": "ordinal",
    },
  ],
  "signals": [
    {
      "name": "plotWidth",
      "value": 85,
    },
    {
      "name": "bandwidth",
      "value": 0.3,
    },
  ],
  "title": {
    "anchor": "start",
    "color": "#000000",
    "dx": 0,
    "font": "sans-serif",
    "fontSize": 20,
    "text": "",
  },
  "width": 700,
}
`;
